### PR TITLE
ENH: Support inplace clip (#15388)

### LINKED ISSUE #15388: clip inplace?
Is it possible to support clip in place, just like numpy does? I assume that would speed up working with large dataframe/series as no extra copies will be needed.

If not, is that a workaround to achieve the same effect? I tried `np.clip(df.values,-clip_bounds.values, clip_bounds.values, out=df.values)`, but realized that df.values is actually an expensive call as it is merging  blocks underneath.

### PR DESCRIPTION
**Pull request by guygoldberg**
- [V] closes #15388
 - [V] tests added / passed
 - [V] passes ``git diff upstream/master --name-only -- '*.py' | flake8 --diff``
 - [x] whatsnew entry

### CODE REVIEW

File: pandas/core/generic.py [hunk 1]
```diff
@@ -4149,6 +4153,8 @@ def clip(self, lower=None, upper=None, axis=None, *args, **kwargs):
         upper : float or array_like, default None
         axis : int or string axis name, optional
             Align object with lower and upper along the given axis.
+        inplace : boolean, default False
+            Whether to perform the operation in place on the data
```
Comment (jreback) [inline pandas/core/generic.py]: needs a versionadded tag.

File: pandas/core/generic.py [hunk 2]
```diff
@@ -4220,6 +4231,8 @@ def clip_upper(self, threshold, axis=None):
         threshold : float or array_like
         axis : int or string axis name, optional
             Align object with threshold along the given axis.
+        inplace : boolean, default False
+            Whether to perform the operation in place on the data
```
Comment (jreback) [inline pandas/core/generic.py]: same

File: pandas/core/generic.py [hunk 3]
```diff
@@ -4247,6 +4262,8 @@ def clip_lower(self, threshold, axis=None):
...
```
Comment (jreback) [inline pandas/core/generic.py]: same

File: pandas/tests/frame/test_analytics.py [hunk 4]
```diff
@@ -1843,18 +1863,25 @@ def test_clip_mixed_numeric(self):
...
         # GH #6966
 
         df = DataFrame(np.random.randn(1000, 2))
         lb = Series(np.random.randn(1000))
         ub = lb + 1
 
-        clipped_df = df.clip(lb, ub, axis=0)
+        original = df.copy()
+
+        if inplace:
```
Comment (jreback) [inline pandas/tests/frame/test_analytics.py]: you can just do

```
clipped_df = df.clip(lb, ub, axis=0, inplace=inplace)

if inplace:
    clipped_df = df
```

**Review by jreback — APPROVED** [review]
minor comment. pls add a whatsnew entry (0.21.0) in enhancements.

**Comment by jreback** [comment]
note this doesn't offer any perf benefit (actually doing this is a bit tricky), but ok.

**Comment by guygoldberg** [comment]
@jreback, thanks. I fixed all of the comments and pushed to my branch.

File: doc/source/whatsnew/v0.21.0.txt [hunk 5]
```diff
@@ -35,6 +35,7 @@ Other Enhancements
 - ``RangeIndex.append`` now returns a ``RangeIndex`` object when possible (:issue:`16212`)
 - :func:`to_pickle` has gained a protocol parameter (:issue:`16252`). By default, this parameter is set to `HIGHEST_PROTOCOL <https://docs.python.org/3/library/pickle.html#data-stream-format>`__
 - :func:`api.types.infer_dtype` now infers decimals. (:issue: `15690`)
+- :func:`clip` now supports inplace trimming. (:issue: `15388`)
```
Comment (jreback) [inline doc/source/whatsnew/v0.21.0.txt]: no, ``DataFrame.clip()`` and ``Series.cip()`` have gained an ``inplace`` argument.

File: pandas/tests/frame/test_analytics.py [hunk 6]
```diff
@@ -1843,18 +1863,25 @@ def test_clip_mixed_numeric(self):
...
         # GH #6966
 
         df = DataFrame(np.random.randn(1000, 2))
         lb = Series(np.random.randn(1000))
         ub = lb + 1
 
-        clipped_df = df.clip(lb, ub, axis=0)
+        original = df.copy()
+
+
```
Comment (jreback) [inline pandas/tests/frame/test_analytics.py]: nit. extra line.

**Review by jreback — APPROVED** [review]
minor comments. needs a rebase on master though. I think you might have some conflicts. ping when green.

**Comment by jreback** [comment]
thanks!

**Pull request merged**